### PR TITLE
3body

### PR DESCRIPTION
**Pull request by schererc**
Additions to the votca package in order to parametrize 3-body interactions of the SW type with csg_fmatch and calculate additional properties with csg_stat

Needs votca/tools#65

### CODE REVIEW

File: src/tools/csg_fmatch.cc [hunk 1]
```diff
@@ -283,39 +372,41 @@ void CGForceMatching::EvalConfiguration(Topology *conf, Topology *conf_atom)
...
         for (int iatom = 0; iatom < _nbeads; ++iatom) {
-            Force = conf->getBead(iatom)->getF();
+            Force = conf->getBead(iatom)->getF();               
             _b(_least_sq_offset + 3 * _nbeads * _frame_counter + iatom) = Force.x();
             _b(_least_sq_offset + 3 * _nbeads * _frame_counter + _nbeads + iatom) = Force.y();
             _b(_least_sq_offset + 3 * _nbeads * _frame_counter + 2 * _nbeads + iatom) = Force.z();
         }
     } else {
-        cerr << "\nERROR in csg_fmatch::EvalConfiguration - No forces in configuration!\n" << endl;
+        cerr << "\nERROR in csg_fmatch::EvalConfiguration - No forces in figuration!\n" << endl;
```
Comment (junghans) [inline src/tools/csg_fmatch.cc]: Huh?

File: src/tools/csg_fmatch.cc [hunk 2]
```diff
@@ -283,39 +372,41 @@ void CGForceMatching::EvalConfiguration(Topology *conf, Topology *conf_atom)
...
+            }
     }
     
     // loop for the forces vector: 
     // hack, chage the Has functions..
     if (conf->getBead(0)->HasF()) {
         vec Force(0., 0., 0.);
         for (int iatom = 0; iatom < _nbeads; ++iatom) {
-            Force = conf->getBead(iatom)->getF();
+            Force = conf->getBead(iatom)->getF();
```
Comment (junghans) [inline src/tools/csg_fmatch.cc]: Why adding white spaces?

File: src/tools/csg_fmatch.cc [hunk 3]
```diff
@@ -352,8 +443,8 @@ void CGForceMatching::EvalConfiguration(Topology *conf, Topology *conf_atom)
 
 void CGForceMatching::FmatchAccumulateData() 
 {
-    if (_constr_least_sq) { // Constrained Least Squares
-        // Solving linear equations system
+    if (_constr_least_sq) { // Constrained Least Squares
```
Comment (junghans) [inline src/tools/csg_fmatch.cc]: Whitespaces?

File: src/tools/csg_stat_imc.h [hunk 4]
```diff
@@ -65,9 +65,18 @@ class Imc
         int _index;
         Property *_p;
         HistogramNew _average;
+        HistogramNew _average_force;
+        /*HistogramNew _average_force_perp;
+        HistogramNew _average_force_perp_dot;        
+        HistogramNew _average_force_perp_x;
+        HistogramNew _average_force_perp_y;
+        HistogramNew _average_force_perp_z;*/
```
Comment (junghans) [inline src/tools/csg_stat_imc.h]: Remove commented out parts!

File: src/tools/csg_stat_imc.cc [hunk 5]
```diff
@@ -136,11 +163,35 @@ Imc::interaction_t *Imc::AddInteraction(Property *p)
...
     int n = static_cast<int>((i->_max - i->_min) / i->_step + 1.000000001);
 
     i->_average.Initialize(i->_min, i->_max, n);
+    if (i->_force){
+        i->_average_force.Initialize(i->_min, i->_max, n);
+        /*i->_average_force_perp.Initialize(i->_min, i->_max, n);
+        i->_average_force_perp_dot.Initialize(i->_min, i->_max, n);        
+        i->_average_force_perp_x.Initialize(i->_min, i->_max, n);
+        i->_average_force_perp_y.Initialize(i->_min, i->_max, n);
+        i->_average_force_perp_z.Initialize(i->_min, i->_max, n);*/
```
Comment (junghans) [inline src/tools/csg_stat_imc.cc]: Remove!

File: src/tools/csg_stat_imc.cc [hunk 6]
```diff
@@ -188,7 +239,15 @@ void Imc::ClearAverages()
...
     for (ic_iter = _interactions.begin(); ic_iter != _interactions.end(); ++ic_iter)
-        ic_iter->second->_average.Clear();    
+        ic_iter->second->_average.Clear();
+        if (ic_iter->second->_force){
+            ic_iter->second->_average_force.Clear();
+            /*ic_iter->second->_average_force_perp.Clear();
+            ic_iter->second->_average_force_perp_dot.Clear();            
+            ic_iter->second->_average_force_perp_x.Clear();
+            ic_iter->second->_average_force_perp_y.Clear();            
+            ic_iter->second->_average_force_perp_z.Clear();*/
```
Comment (junghans) [inline src/tools/csg_stat_imc.cc]: Remove!

**Review by junghans — CHANGES_REQUESTED** [review]
Needs a test! See https://github.com/votca/csg/blob/master/src/tools/CMakeLists.txt#L38 for an example.

File: src/tools/csg_fmatch.cc (hunk 1 above)
Comment (schererc) [reply src/tools/csg_fmatch.cc]: Okay, that is a mistake

File: src/tools/csg_fmatch.cc (hunk 2 above)
Comment (schererc) [reply src/tools/csg_fmatch.cc]: That probably happened due to automerge

File: src/tools/csg_fmatch.cc (hunk 3 above)
Comment (schererc) [reply src/tools/csg_fmatch.cc]: Again, that probably happened due to automerge

File: src/tools/csg_stat_imc.h (hunk 4 above)
Comment (schererc) [reply src/tools/csg_stat_imc.h]: Okay

File: src/tools/csg_stat_imc.cc (hunk 5 above)
Comment (schererc) [reply src/tools/csg_stat_imc.cc]: Okay

File: src/tools/csg_stat_imc.cc (hunk 6 above)
Comment (schererc) [reply src/tools/csg_stat_imc.cc]: Okay

File: include/votca/csg/nblist_3body.h [hunk 7]
```diff
@@ -0,0 +1,175 @@
+/* 
+ * Copyright 2009-2016 The VOTCA Development Team (http://www.votca.org)
```
Comment (JoshuaSBrown) [inline include/votca/csg/nblist_3body.h]: Update date?

File: include/votca/csg/nblist_3body.h [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#ifndef _NBLIST_3BODY_H
+#define	_NBLIST_3BODY_H
+
```
Comment (JoshuaSBrown) [inline include/votca/csg/nblist_3body.h:20]: Change to _VOTCA_CSG_NBLIST_3BODY_H

File: include/votca/csg/nblist_3body.h [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+#include "beadlist.h"
+#include "beadtriple.h"
+#include "triplelist.h"
+#include "exclusionlist.h"
+
+namespace votca { namespace csg {
+using namespace votca::tools;
+
```
Comment (JoshuaSBrown) [inline include/votca/csg/nblist_3body.h:27]: If possible remove using namespace votca::tools and use full scope in header file.

File: include/votca/csg/nblist_3body.h [hunk 10]
```diff
@@ -0,0 +1,175 @@
...
+inline void NBList_3Body::SetMatchFunction(bool (*fkt)(Bead *, Bead *, Bead *, const vec &, const vec &, const vec &, const double dist12, const double dist13, const double dist23))
+{
+    if(_match_function)
+        delete _match_function;
+    _match_function = dynamic_cast<Functor*>(new FunctorNonMember(fkt));
+}
+
+}}
+
+#endif	/* _NBLIST_3BODY_H */
```
Comment (JoshuaSBrown) [inline include/votca/csg/nblist_3body.h]: change to 
// _VOTCA_CSG_NBLIST_3BODY_H

File: include/votca/csg/nblistgrid.h [hunk 11]
```diff
@@ -1,5 +1,5 @@
 /* 
- * Copyright 2009-2011 The VOTCA Development Team (http://www.votca.org)
+ * Copyright 2009-2016 The VOTCA Development Team (http://www.votca.org)
```
Comment (JoshuaSBrown) [inline include/votca/csg/nblistgrid.h]: Was this file updated?

File: include/votca/csg/nblistgrid_3body.h [hunk 12]
```diff
@@ -0,0 +1,68 @@
+/* 
+ * Copyright 2009-2017 The VOTCA Development Team (http://www.votca.org)
```
Comment (JoshuaSBrown) [inline include/votca/csg/nblistgrid_3body.h]: Update?

File: include/votca/csg/nblistgrid_3body.h [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#ifndef NBLISTGRID_3BODY_H
+#define	NBLISTGRID_3BODY_H
+
```
Comment (JoshuaSBrown) [inline include/votca/csg/nblistgrid_3body.h:20]: Change to VOTCA_CSG_NBLISTGRID_3BODY_H

File: include/votca/csg/nblistgrid_3body.h [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+#include <votca/tools/matrix.h>
+#include <votca/tools/vec.h>
+#include "nblist_3body.h"
+#include <vector>
+
+namespace votca { namespace csg {
+using namespace votca::tools;
```
Comment (JoshuaSBrown) [inline include/votca/csg/nblistgrid_3body.h]: Remove using namespace from header file, I suggest using full scope or the abbreviation
```
namespace TOOLS = votca::tools;

File: include/votca/csg/nblistgrid_3body.h [hunk 15]
```diff
@@ -0,0 +1,68 @@
...
+class NBListGrid_3Body
+    : public NBList_3Body
+{
+public:
+    void Generate(BeadList &list1, BeadList &list2, BeadList &list3, bool do_exclusions = true);
+    void Generate(BeadList &list1, BeadList &list2, bool do_exclusions = true);
+    void Generate(BeadList &list, bool do_exclusions = true);
+
+protected:
+    struct cell_t {
```
Comment (JoshuaSBrown) [inline include/votca/csg/nblistgrid_3body.h:37]: Could you give this a more descriptive name?

File: include/votca/csg/triplelist.h [hunk 16]
```diff
@@ -58,10 +59,22 @@ class TripleList {
...
+    //experimental: be careful, if same triple object is used, some values might change (e.g. sign of distance vectors)
+    //(*t)[i] gives access to ith element of tuple object (i=0,1,2).
+    //there are 6 possible permutations of e1,e2,e3 -> still the same triple of elements
+    /*_triple_map[ (*t)[0] ][ (*t)[1] ][ (*t)[2] ] = t;
+    _triple_map[ (*t)[0] ][ (*t)[2] ][ (*t)[1] ] = t;
+    _triple_map[ (*t)[1] ][ (*t)[0] ][ (*t)[2] ] = t;
+    _triple_map[ (*t)[1] ][ (*t)[2] ][ (*t)[0] ] = t;
+    _triple_map[ (*t)[2] ][ (*t)[0] ][ (*t)[1] ] = t;
+    _triple_map[ (*t)[2] ][ (*t)[1] ][ (*t)[0] ] = t;*/
+
```
Comment (JoshuaSBrown) [inline include/votca/csg/triplelist.h]: Is this commented out code or is it part of the comment?

File: src/libcsg/nblist_3body.cc [hunk 17]
```diff
@@ -0,0 +1,116 @@
+/* 
+ * Copyright 2009-2011 The VOTCA Development Team (http://www.votca.org)
```
Comment (JoshuaSBrown) [inline src/libcsg/nblist_3body.cc:2]: date?

File: src/libcsg/nblistgrid_3body.cc [hunk 18]
```diff
@@ -0,0 +1,264 @@
+/* 
+ * Copyright 2009-2017 The VOTCA Development Team (http://www.votca.org)
```
Comment (JoshuaSBrown) [inline src/libcsg/nblistgrid_3body.cc:2]: date?

File: src/tools/csg_stat_imc.h [hunk 19]
```diff
@@ -136,6 +140,12 @@ class Imc
     public:
 
         vector<HistogramNew> _current_hists;
+        vector<HistogramNew> _current_hists_force;
+        /*vector<HistogramNew> _current_hists_force_perp;
```
Comment (JoshuaSBrown) [inline src/tools/csg_stat_imc.h]: Just delete?

**Review by JoshuaSBrown — CHANGES_REQUESTED** [review]
Optional:
Use clang-format to apply google rules to the files you have changed

Required:
Add testing of the classes and methods you have implemented
Remove any using namespace calls from header files. 

For my sake could you also provide a dumbed down explanation of what the code is meant to be doing and how it fits in with the rest of VOTCA, maybe we could communicate via slack? It would give me a better idea of what I'm looking at. As a side note, the classes and functions are very cleanly implemented 👍 very nice. 

Also just in case you were not aware there is a developers guide here:
https://github.com/votca/votca/blob/master/share/doc/DEVELOPERS_GUIDE.md#cpp-codeing-style-guide

File: include/votca/csg/nblist_3body.h (hunk 7 above)
Comment (schererc) [reply include/votca/csg/nblist_3body.h]: I now have changed this

File: include/votca/csg/nblist_3body.h (hunk 8 above)
Comment (schererc) [reply include/votca/csg/nblist_3body.h:20]: I changed this here and in all the other header files

File: include/votca/csg/nblist_3body.h (hunk 9 above)
Comment (schererc) [reply include/votca/csg/nblist_3body.h:27]: I removed the using namespace here and in the other newly added header files

File: include/votca/csg/nblist_3body.h (hunk 10 above)
Comment (schererc) [reply include/votca/csg/nblist_3body.h]: changed

File: include/votca/csg/nblistgrid.h (hunk 11 above)
Comment (schererc) [reply include/votca/csg/nblistgrid.h]: changed

File: include/votca/csg/triplelist.h (hunk 16 above)
Comment (schererc) [reply include/votca/csg/triplelist.h]: It is commented out code that I removed

File: src/libcsg/nblist_3body.cc (hunk 17 above)
Comment (schererc) [reply src/libcsg/nblist_3body.cc:2]: yes

File: src/libcsg/nblistgrid_3body.cc (hunk 18 above)
Comment (schererc) [reply src/libcsg/nblistgrid_3body.cc:2]: changed

File: src/tools/csg_stat_imc.h (hunk 19 above)
Comment (schererc) [reply src/tools/csg_stat_imc.h]: I deleted it

File: include/votca/csg/beadtriple.h [hunk 20]
```diff
@@ -41,6 +40,11 @@ class BeadTriple
         
     virtual ~BeadTriple() {}
 
+    /// \brief return the beads
+    Bead * bead1() { return std::get<0>(*this); }
+    Bead * bead2() { return std::get<1>(*this); }
+    Bead * bead3() { return std::get<2>(*this); }
+    
     /// \brief the vector connecting two beads
     vec &r12() { return _r12; }
```
Comment (JensWehner) [inline include/votca/csg/beadtriple.h:49]: I also would add `const ` accessors

File: include/votca/csg/nblist_3body.h [hunk 21]
```diff
@@ -0,0 +1,174 @@
...
+    /// Experimental: here the second and the third bead list are the same (nblist will have the structure bead type 1, bead type 2, bead type 2) 
+    virtual void Generate(BeadList &list1, BeadList &list2, bool do_exclusions = true) { Generate(list1, list2, list2, do_exclusions); };
+    /// Generate the 3body neighbour list based on a single bead list (nblist will have the structure bead type 1, bead type 1, bead type 1)
+    virtual void Generate(BeadList &list, bool do_exclusions = true) { Generate(list, list, list, do_exclusions); }
+    
+    /// set the cutoff for the neighbour search
+    /// to do: at the moment use only one single cutoff value
+    void setCutoff(double cutoff) { _cutoff = cutoff; }
+    /// get the cutoff for the neighbour search
+    double getCutoff() { return _cutoff; }
```
Comment (JensWehner) [inline include/votca/csg/nblist_3body.h]: can be declared `const`

File: include/votca/csg/nblist_3body.h [hunk 22]
```diff
@@ -0,0 +1,174 @@
...
+ */
+class NBList_3Body 
+    : public TripleList<Bead*, BeadTriple>
+{
+public:
+    NBList_3Body();
+    virtual ~NBList_3Body();    
+
+    /// Generate the 3body neighbour list based on three bead lists (e.g. bead types)
+    virtual void Generate(BeadList &list1, BeadList &list2, BeadList &list3, bool do_exclusions = true);
```
Comment (JensWehner) [inline include/votca/csg/nblist_3body.h:48]: why are all these functions virtual, do you want to inherit form NBList_3Body? And if so, why not make them pure virtual functions

File: include/votca/csg/triplelist.h [hunk 23]
```diff
@@ -32,24 +31,25 @@ class TripleList {
...
     
     void Cleanup();
     
     triple_type *FindTriple(element_type e1, element_type e2, element_type e3);
+    
 
     typedef element_type element_t;
     typedef triple_type triple_t;
 
 protected:
```
Comment (JensWehner) [inline include/votca/csg/triplelist.h]: why not private, but protected?

File: include/votca/csg/triplelist.h [hunk 24]
```diff
@@ -58,10 +58,12 @@ class TripleList {
 template<typename element_type, typename triple_type>
 inline void TripleList<element_type, triple_type>::AddTriple(triple_type *t)
 {
-    /// \todo be careful, same triple object is used, some values might change (e.g. sign of distance vectors)
-    //experimental: So far only mapping '123' and '321' to the same triple
-    _triple_map[ (*t)[0] ][ (*t)[1] ][ (*t)[2] ] = t;
-    _triple_map[ (*t)[2] ][ (*t)[1] ][ (*t)[0] ] = t;
+    //experimental: be careful, if same triple object is used, some values might change (e.g. sign of distance vectors)
```
Comment (JensWehner) [inline include/votca/csg/triplelist.h]: If it is an experimental feature it should not be in master. I do not understand the comment

File: src/libcsg/nblist_3body.cc [hunk 25]
```diff
@@ -0,0 +1,116 @@
...
+NBList_3Body::NBList_3Body()
+  : _do_exclusions(false), _match_function(0)
+{
+    setTripleType<BeadTriple>();
+    SetMatchFunction(NBList_3Body::match_always);    
+}
+
+NBList_3Body::~NBList_3Body()
+{
+  // TODO: NBList_3Body destructor
```
Comment (JensWehner) [inline src/libcsg/nblist_3body.cc]: I am not sure, if it only has std:: datatypes the default destructor is enough, if it has not it is a memory leak and it should not be in master

File: src/libcsg/nblistgrid_3body.cc [hunk 26]
```diff
@@ -0,0 +1,264 @@
...
+    //check if all bead lists "have" the same topology
+    assert(list1.getTopology() == list2.getTopology());
+    assert(list1.getTopology() == list3.getTopology());
+    assert(list2.getTopology() == list3.getTopology());
+    Topology *top = _top = list1.getTopology();
+
+    InitializeGrid(top->getBox());
+
+    // Add all beads of list1 to _beads1
+    for(iter = list1.begin(); iter != list1.end(); ++iter)
```
Comment (JensWehner) [inline src/libcsg/nblistgrid_3body.cc:39]: okay, but C++11 range loops are nicer

Comment (JensWehner) [reply src/libcsg/nblistgrid_3body.cc:39]:
```
for(auto bead*:list1){
getCell(bead->getPos()
}
```

File: src/libcsg/nblistgrid_3body.cc [hunk 27]
```diff
@@ -0,0 +1,264 @@
...
+    
+}
+    
+void NBListGrid_3Body::Generate(BeadList &list1, BeadList &list2, bool do_exclusions)
+{
+    BeadList::iterator iter;
+    _do_exclusions = do_exclusions;
+    if(list1.empty()) return;
+    if(list2.empty()) return;
+
```
Comment (JensWehner) [inline src/libcsg/nblistgrid_3body.cc:64]: This is really similar to the code above can you not move it out into a single function?

File: src/libcsg/nblistgrid_3body.cc [hunk 28]
```diff
@@ -0,0 +1,264 @@
...
+    Topology *top = _top = list.getTopology();
+
+    InitializeGrid(top->getBox());
+    
+    // Add all beads of list to all! bead lists of the cell
+    for(iter = list.begin(); iter != list.end(); ++iter) 
+        getCell((*iter)->getPos())._beads1.push_back(*iter);    
+    
+    for(vector<cell_t>::iterator iter = _grid.begin(); iter != _grid.end(); ++iter){
```
Comment (JensWehner) [inline src/libcsg/nblistgrid_3body.cc:105]: C++11 for loops are muhc nicer

File: src/libcsg/nblistgrid_3body.cc [hunk 29]
```diff
@@ -0,0 +1,264 @@
...
+    for(int a=_box_Na; a<2*_box_Na; ++a)
+        for(int b=_box_Nb; b<2*_box_Nb; ++b)
+            for(int c=_box_Nc; c<2*_box_Nc; ++c) {
+                cell_t &cell = getCell(a%_box_Na, b%_box_Nb, c%_box_Nc);
+                for(int aa=a+a1; aa<=a+a2; ++aa)
+                    for(int bb=b+b1; bb<=b+b2; ++bb)
+                        for(int cc=c+c1; cc<=c+c2; ++cc) {
+                            cell_t *c = &getCell(aa%_box_Na, bb%_box_Nb, cc%_box_Nc);
+                            //test: for 3body algorithm: each cell is a neighbor of its own !!!
+                            //if(c == &cell) continue; // ignore self
```
Comment (JensWehner) [inline src/libcsg/nblistgrid_3body.cc]: remove commented out code

File: src/libcsg/nblistgrid_3body.cc [hunk 30]
```diff
@@ -0,0 +1,264 @@
...
+
+
+void NBListGrid_3Body::TestBead(NBListGrid_3Body::cell_t &cell, Bead *bead)
+{
+    BeadList::iterator iter2;
+    BeadList::iterator iter3;    
+    vec u = bead->getPos();
+    
+    //loop over all neighbors (this now includes the cell itself!) to iterate over all beads of type2 of the cell and its neighbors
+    for(vector<cell_t*>::iterator iterc2=cell._neighbours.begin(); iterc2!=cell._neighbours.end(); ++iterc2) {
```
Comment (JensWehner) [inline src/libcsg/nblistgrid_3body.cc:205]: C++11 for loop

File: src/tools/csg_fmatch.cc [hunk 31]
```diff
@@ -537,3 +638,138 @@ void CGForceMatching::EvalNonbonded(Topology *conf, SplineInfo *sinfo)
...
+        vec rij  = (*triple_iter)->r12();
+        vec rik  = (*triple_iter)->r13();
+        
+        double gamma_sigma = (sinfo->gamma)*(sinfo->sigma);
+        double denomij = (distij-(sinfo->a)*(sinfo->sigma));
+        double denomik = (distik-(sinfo->a)*(sinfo->sigma));
+        double expij = exp(gamma_sigma/denomij);
+        double expik = exp(gamma_sigma/denomik);
+        
+        //std::cout << "iatom: " << iatom << " , jatom: " << jatom << " , katom: " << katom << std::endl;
```
Comment (JensWehner) [inline src/tools/csg_fmatch.cc]: remove

File: src/tools/csg_fmatch.cc [hunk 32]
```diff
@@ -537,3 +638,138 @@ void CGForceMatching::EvalNonbonded(Topology *conf, SplineInfo *sinfo)
...
+            //type1 != type2 != type3
+            if (sinfo->type1 != sinfo->type3){              
+                nb->Generate(beads1, beads2, beads3, true);
+            }
+        }
+    }
+    
+    NBList_3Body::iterator triple_iter;
+    // iterate over all triples
+    for (triple_iter = nb->begin(); triple_iter != nb->end(); ++triple_iter){
```
Comment (JensWehner) [inline src/tools/csg_fmatch.cc:704]: C++11 for loops

File: src/tools/csg_fmatch.cc [hunk 33]
```diff
@@ -537,3 +638,138 @@ void CGForceMatching::EvalNonbonded(Topology *conf, SplineInfo *sinfo)
...
+        CubicSpline &SP = sinfo->Spline;
+
+        int &mpos = sinfo->matr_pos;
+        
+        double var = acos(rij*rik/sqrt((rij*rij) * (rik*rik)));
+        
+        double acos_prime = 1.0 / (sqrt(1 - (rij*rik) * (rij*rik)/( distij * distik * distij * distik ) ));
+        
+        //evaluate gradient1 and gradient2 for iatom:        
+        //gradient1 = (-grad_i theta) * exp()*exp()
```
Comment (JensWehner) [inline src/tools/csg_fmatch.cc]: remove commented out code

File: src/tools/csg_fmatch.cc [hunk 34]
```diff
@@ -537,3 +638,138 @@ void CGForceMatching::EvalNonbonded(Topology *conf, SplineInfo *sinfo)
...
+        // add iatom
+        SP.AddToFitMatrix(_A, var,
+                _least_sq_offset + 3 * _nbeads * _frame_counter + iatom, mpos, -gradient1.x(), -gradient2.x());
+        SP.AddToFitMatrix(_A, var,
+                _least_sq_offset + 3 * _nbeads * _frame_counter + _nbeads + iatom, mpos, -gradient1.y(), -gradient2.y());
+        SP.AddToFitMatrix(_A, var,
+                _least_sq_offset + 3 * _nbeads * _frame_counter + 2 * _nbeads + iatom, mpos, -gradient1.z(), -gradient2.z());
+        
+        //evaluate gradient1 and gradient2 for jatom:        
+        //gradient1 = (-grad_j theta) * exp()*exp()
```
Comment (JensWehner) [inline src/tools/csg_fmatch.cc]: again

File: src/tools/csg_fmatch.cc [hunk 35]
```diff
@@ -537,3 +638,138 @@ void CGForceMatching::EvalNonbonded(Topology *conf, SplineInfo *sinfo)
...
+        // add jatom
+        SP.AddToFitMatrix(_A, var,
+                _least_sq_offset + 3 * _nbeads * _frame_counter + jatom, mpos, -gradient1.x(), -gradient2.x());
+        SP.AddToFitMatrix(_A, var,
+                _least_sq_offset + 3 * _nbeads * _frame_counter + _nbeads + jatom, mpos, -gradient1.y(), -gradient2.y());
+        SP.AddToFitMatrix(_A, var,
+                _least_sq_offset + 3 * _nbeads * _frame_counter + 2 * _nbeads + jatom, mpos, -gradient1.z(), -gradient2.z());
+        
+        //evaluate gradient1 and gradient2 for katom:        
+        //gradient1 = (-grad_k theta) * exp()*exp()
```
Comment (JensWehner) [inline src/tools/csg_fmatch.cc]: again

File: src/tools/csg_stat_imc.cc [hunk 36]
```diff
@@ -60,7 +61,7 @@ void Imc::Initialize()
         i->_is_bonded = false;
     }
 
-    // initialize non-bonded structures
+    // initialize bonded structures
    for (list<Property*>::iterator iter = _bonded.begin();
```
Comment (JensWehner) [inline src/tools/csg_stat_imc.cc:65]: C++11 loop

File: src/tools/csg_stat_imc.cc [hunk 37]
```diff
@@ -136,11 +163,30 @@ Imc::interaction_t *Imc::AddInteraction(Property *p)
...
+    //check if option threebody exists
+    if (p->exists("threebody")) {
+        i->_threebody = p->get("threebody").as<bool>();            
+    }
+    //check if option force exists
+    if (p->exists("force")) {
+        i->_force = p->get("force").as<bool>();            
+    }    
+    //preliminary. should be changed
+    i->_cut = 0.37; //(0.37 nm)
```
Comment (JensWehner) [inline src/tools/csg_stat_imc.cc]: tools has now a function ```ifExistsReturnElseReturnDefault``` which does all this a bit shorter

File: src/tools/csg_stat_imc.cc [hunk 38]
```diff
@@ -235,34 +279,140 @@ void Imc::Worker::DoNonbonded(Topology *top)
...
-        delete nb;
+            // is it same types or different types?
+            if((*iter)->get("type1").value() == (*iter)->get("type2").value())
+                nb->Generate(beads1);
+            else
+                nb->Generate(beads1, beads2);                      
+                                     
+            
+            // process all pairs
+            /*NBList::iterator pair_iter;
```
Comment (JensWehner) [inline src/tools/csg_stat_imc.cc]: delete commented out code

File: src/tools/csg_stat_imc.cc [hunk 39]
```diff
@@ -379,30 +529,73 @@ void Imc::WriteDist(const string &suffix)
         // calculate the rdf
         Table &t = iter->second->_average.data();            
         Table dist(t);
+        
+        //preliminary, if no average force calculation, dummy table
+        Table force;
+        //Table force_perp,force_perp_dot,force_perp_x,force_perp_y,force_perp_z;
```
Comment (JensWehner) [inline src/tools/csg_stat_imc.cc]: again

**Review by JensWehner — COMMENTED** [review]
I like the unittests so in general good idea. I highlighted a couple of things to improve code quality as I am not a maintainer of the code, I will not request changes

File: include/votca/csg/beadtriple.h (hunk 20 above)
Comment (schererc) [reply include/votca/csg/beadtriple.h:49]: Okay, I changed it

File: include/votca/csg/nblist_3body.h (hunk 21 above)
Comment (schererc) [reply include/votca/csg/nblist_3body.h]: I changed it

File: include/votca/csg/nblist_3body.h (hunk 22 above)
Comment (schererc) [reply include/votca/csg/nblist_3body.h:48]: Actually that is what happens. nblistgrid_3body.h inherits from nblist_3body.h

File: include/votca/csg/beadtriple.h (hunk 20 above)
Comment (JensWehner) [reply include/votca/csg/beadtriple.h:49]: I am not sure if you might need both. It is more a proposal than a you have too

File: include/votca/csg/triplelist.h (hunk 23 above)
Comment (schererc) [reply include/votca/csg/triplelist.h]: changed it

File: include/votca/csg/triplelist.h (hunk 24 above)
Comment (schererc) [reply include/votca/csg/triplelist.h]: That was an old comment that was still in the code, removed it

File: src/libcsg/nblist_3body.cc (hunk 25 above)
Comment (schererc) [reply src/libcsg/nblist_3body.cc]: No, it does not have a memory leak. I removed the todo statement

File: src/libcsg/nblistgrid_3body.cc (hunk 26 above)
Comment (schererc) [reply src/libcsg/nblistgrid_3body.cc:39]: okay, that's true. But maybe one should replace all loops of this type by C++11 range loops at once everywhere in the votca package

File: src/libcsg/nblistgrid_3body.cc (hunk 28 above)
Comment (schererc) [reply src/libcsg/nblistgrid_3body.cc:105]: Okay, next time!

File: src/libcsg/nblistgrid_3body.cc (hunk 29 above)
Comment (schererc) [reply src/libcsg/nblistgrid_3body.cc]: okay

File: src/tools/csg_fmatch.cc (hunk 32 above)
Comment (schererc) [reply src/tools/csg_fmatch.cc:704]: next time

File: src/libcsg/nblistgrid_3body.cc (hunk 27 above)
Comment (schererc) [reply src/libcsg/nblistgrid_3body.cc:64]: Okay, they are similar. But, actually these are three different cases. Either all three bead types are the same, only 2 bead types are the same or all bead types are different.

File: src/libcsg/nblistgrid_3body.cc (hunk 30 above)
Comment (schererc) [reply src/libcsg/nblistgrid_3body.cc:205]: okay

Comment (schererc) [reply src/libcsg/nblistgrid_3body.cc:205]: I would do it as a general cleanup of the code

File: src/tools/csg_fmatch.cc (hunk 31 above)
Comment (schererc) [reply src/tools/csg_fmatch.cc]: changed it

File: src/tools/csg_fmatch.cc (hunk 33 above)
Comment (schererc) [reply src/tools/csg_fmatch.cc]: changed it

File: src/tools/csg_fmatch.cc (hunk 34 above)
Comment (schererc) [reply src/tools/csg_fmatch.cc]: changed it

File: src/tools/csg_fmatch.cc (hunk 35 above)
Comment (schererc) [reply src/tools/csg_fmatch.cc]: changed it

File: src/tools/csg_stat_imc.cc (hunk 36 above)
Comment (schererc) [reply src/tools/csg_stat_imc.cc:65]: okay, next time

File: src/tools/csg_stat_imc.cc (hunk 38 above)
Comment (schererc) [reply src/tools/csg_stat_imc.cc]: changed it

File: src/tools/csg_stat_imc.cc (hunk 39 above)
Comment (schererc) [reply src/tools/csg_stat_imc.cc]: changed it

File: src/tools/csg_stat_imc.cc (hunk 37 above)
Comment (schererc) [reply src/tools/csg_stat_imc.cc]: changed

**Comment by JensWehner** [comment]
job is failing quite badly

https://api.travis-ci.org/v3/job/458693975/log.txt

schererc: I looked again into the log file. The build is failing because the added function in cubicspline.h of votca/tools is not yet in the master branch (pending pull request of votca/tools) and so travis does not find it yet. So when the 3body branch of votca/tools is merged into the master branch everything should work! (At least it does for me locally).

**Comment by JensWehner** [comment]
You can just comment below.and do not have to edit my comment :) There was also a second error 

https://api.travis-ci.org/v3/job/458693979/log.txt

**Comment by junghans** [comment]
Travis is failing as we need to merge votca/tools#65 first

File: src/libcsg/nblistgrid_3body.cc (hunk 18 above)
Comment (junghans) [reply src/libcsg/nblistgrid_3body.cc:2]: 2018?

**Review by junghans — CHANGES_REQUESTED** [review]
Can you add a csg_stat regression testing as well? (See https://github.com/votca/csg/blob/master/src/tools/CMakeLists.txt#L71)

**Comment by schererc** [comment]
I have now added regression tests for csg_stat for the 2-body potential of mean force, and the 3-body angular distribution, as well as, csg_fmatch for the 2-body + 3-body SW interaction based on the frame.dump file in src/tools/references/spce

**Comment by junghans** [comment]
@schererc merge https://github.com/votca/votca/pull/94 then this should build.

**Comment by junghans** [comment]
```
32/56 Test #55: unit_test_nblist_3body .............***Failed    0.02 sec
unit_test_nblist_3body: /home/votca/votca/csg/include/votca/csg/moleculeitem.h:36: votca::csg::Molecule* votca::csg::MoleculeItem::getMolecule(): Assertion `_mol != nullptr' failed.
Running 2 test cases...
unknown location(0): fatal error: in "nblist_3body_test/test_nblist_3body_generate_list": signal: SIGABRT (application abort requested)
/home/votca/votca/csg/src/tests/test_nblist_3body.cc(41): last checkpoint: "test_nblist_3body_generate_list" test entry
*** 1 failure is detected in the test module "nblist_3body_test"
      Start 56: unit_test_nblistgrid_3body
33/56 Test #56: unit_test_nblistgrid_3body .........***Failed    0.02 sec
```
it seems the assertion in `/home/votca/votca/csg/include/votca/csg/moleculeitem.h:36` is failing in the debug build.

**Comment by junghans** [comment]
@schererc you can do a debug build locally by adding `-DCMAKE_BUILD_TYPE=Debug` to your `cmake` options.

**Comment by schererc** [comment]
I assume, I found and corrected now the problems in test_nblist_3body.cc and test_nblistgrid_3body.cc

**Review by junghans — APPROVED** [review]

**Comment by junghans** [comment]
Ping @JensWehner @JoshuaSBrown

**Review by JoshuaSBrown — APPROVED** [review]

**Comment by schererc** [comment]
Thanks, I will then merge the pull request now

**Pull request merged**